File: FitForge/Models/WorkoutSession.swift
import Foundation
import SwiftData

/// 单次训练记录
@Model
final class WorkoutSession {
    var id: UUID
    var date: Date
    var dayType: WorkoutDayType
    var durationMinutes: Int
    var notes: String
    var isCompleted: Bool
    var caloriesBurned: Int

    @Relationship(deleteRule: .cascade)
    var exerciseRecords: [ExerciseRecord]

    init(
        date: Date = Date(),
        dayType: WorkoutDayType,
        durationMinutes: Int = 0,
        notes: String = "",
        isCompleted: Bool = false,
        caloriesBurned: Int = 0,
        exerciseRecords: [ExerciseRecord] = []
    ) {
        self.id = UUID()
        self.date = date
        self.dayType = dayType
        self.durationMinutes = durationMinutes
        self.notes = notes
        self.isCompleted = isCompleted
        self.caloriesBurned = caloriesBurned
        self.exerciseRecords = exerciseRecords
    }
}

/// 单个动作的训练记录
@Model
final class ExerciseRecord {
    var id: UUID
    var exerciseId: UUID
    var exerciseName: String
    var date: Date

    @Relationship(deleteRule: .cascade)
    var sets: [SetRecord]

    @Relationship(inverse: \WorkoutSession.exerciseRecords)
    var session: WorkoutSession?

    init(
        exerciseId: UUID,
        exerciseName: String,
        date: Date = Date(),
        sets: [SetRecord] = []
    ) {
        self.id = UUID()
        self.exerciseId = exerciseId
        self.exerciseName = exerciseName
        self.date = date
        self.sets = sets
    }

    /// 该动作的总容量 (重量 × 次数 之和)
    var totalVolume: Double {
        sets.reduce(0) { $0 + $1.weightKg * Double($1.reps) }
    }
}

/// 单组记录
@Model
final class SetRecord {
    var id: UUID
    var setNumber: Int
    var weightKg: Double
    var reps: Int
    var isCompleted: Bool
    var rpe: Double? // 自觉疲劳度 (Rate of Perceived Exertion, 1-10)

    @Relationship(inverse: \ExerciseRecord.sets)
    var exerciseRecord: ExerciseRecord?

    init(
        setNumber: Int,
        weightKg: Double = 0,
        reps: Int = 0,
        isCompleted: Bool = false,
        rpe: Double? = nil
    ) {
        self.id = UUID()
        self.setNumber = setNumber
        self.weightKg = weightKg
        self.reps = reps
        self.isCompleted = isCompleted
        self.rpe = rpe
    }
}

File: lib/models/workout_session.dart
import 'enums.dart';

/// 单次训练记录
class WorkoutSession {
  final String id;
  final DateTime date;
  final WorkoutDayType dayType;
  int durationMinutes;
  bool isCompleted;
  final List<ExerciseRecord> exerciseRecords;

  WorkoutSession({
    required this.id,
    DateTime? date,
    required this.dayType,
    this.durationMinutes = 0,
    this.isCompleted = false,
    List<ExerciseRecord>? exerciseRecords,
  })  : date = date ?? DateTime.now(),
        exerciseRecords = exerciseRecords ?? [];

  Map<String, dynamic> toJson() => {
        'id': id,
        'date': date.toIso8601String(),
        'dayType': dayType.name,
        'durationMinutes': durationMinutes,
        'isCompleted': isCompleted,
        'exerciseRecords': exerciseRecords.map((r) => r.toJson()).toList(),
      };

  factory WorkoutSession.fromJson(Map<String, dynamic> json) => WorkoutSession(
        id: json['id'] as String,
        date: DateTime.parse(json['date'] as String),
        dayType: WorkoutDayType.values.byName(json['dayType'] as String),
        durationMinutes: json['durationMinutes'] as int,
        isCompleted: json['isCompleted'] as bool,
        exerciseRecords: (json['exerciseRecords'] as List)
            .map((r) => ExerciseRecord.fromJson(r))
            .toList(),
      );
}

/// 单个动作的训练记录
class ExerciseRecord {
  final String exerciseId;
  final String exerciseName;
  final List<SetRecord> sets;

  ExerciseRecord({
    required this.exerciseId,
    required this.exerciseName,
    List<SetRecord>? sets,
  }) : sets = sets ?? [];

  double get totalVolume =>
      sets.fold(0, (sum, s) => sum + s.weightKg * s.reps);

  Map<String, dynamic> toJson() => {
        'exerciseId': exerciseId,
        'exerciseName': exerciseName,
        'sets': sets.map((s) => s.toJson()).toList(),
      };

  factory ExerciseRecord.fromJson(Map<String, dynamic> json) => ExerciseRecord(
        exerciseId: json['exerciseId'] as String,
        exerciseName: json['exerciseName'] as String,
        sets: (json['sets'] as List).map((s) => SetRecord.fromJson(s)).toList(),
      );
}

/// 单组记录
class SetRecord {
  final int setNumber;
  double weightKg;
  int reps;
  bool isCompleted;

  SetRecord({
    required this.setNumber,
    this.weightKg = 0,
    this.reps = 0,
    this.isCompleted = false,
  });

  Map<String, dynamic> toJson() => {
        'setNumber': setNumber,
        'weightKg': weightKg,
        'reps': reps,
        'isCompleted': isCompleted,
      };

  factory SetRecord.fromJson(Map<String, dynamic> json) => SetRecord(
        setNumber: json['setNumber'] as int,
        weightKg: (json['weightKg'] as num).toDouble(),
        reps: json['reps'] as int,
        isCompleted: json['isCompleted'] as bool,
      );
}
